object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java

```diff
@@ -345,13 +378,68 @@ private boolean isTransferOwnershipAtReturn(ControlFlowGraph cfg) {
   }
 
   private void handleAssignment(AssignmentNode node, Set<ImmutableSet<LocalVarWithTree>> newDefs) {
-    Node lhs = node.getTarget();
-    Node rhs = node.getExpression();
+    Node rhs = removeCasts(node.getExpression());
 
-    while (rhs instanceof TypeCastNode) {
-      rhs = ((TypeCastNode) rhs).getOperand();
+    if (rhs instanceof TernaryExpressionNode) {
+      recurseThroughTernaryExpr(node, newDefs, (TernaryExpressionNode) rhs);
+    } else {
+      handleAssignFromRHS(node, newDefs, rhs);
     }
+  }
 
+  /**
+   * Recursively find all method calls nested under a ternary expression on the right-hand side of
+   * an assignment, and handle each as if it were directly assigned to the left-hand side variable.
+   *
+   * @param node original assignment
+   * @param newDefs currently-tracked definitions
+   * @param rhsTernary
+   */
+  private void recurseThroughTernaryExpr(
+      AssignmentNode node,
+      Set<ImmutableSet<LocalVarWithTree>> newDefs,
+      TernaryExpressionNode rhsTernary) {
+    // If handling the then operand of the ternary expression results in adding the assignment to
+    // newDefs, we do not want to also handle the else operand.  Otherwise, for an assignment like
+    // x = b ? new Foo() : new Foo();, the algorithm would add the assignment for the then case, but
+    // then report a false positive error for the else case, since based on the updated state of
+    // newDefs it looks like a tracked variable x is being overwritten.  To avoid these false
+    // positives, we track the size of newDefs before and after handling the then operand, and only
+    // recurse to the else operand if the size does not change.
```

Comment:
I'm concerned about this, because it's possible that the then and else branches will have different, non-empty must-call obligations. Consider e.g.:

```
@MustCall("toString") class Foo extends Object { }
@MustCall("clone") class Bar extends Object { }

static void test_ternary(boolean b) {
    // :: error: required.method.not.called
    Object obj = b ? new Foo() : new Bar();
    // at this point, for soundness, we should be responsible for calling both toString and clone on obj...
    obj.toString();
}
```

Hopefully I'm wrong, but I'm concerned that this code won't handle this correctly, because it'll add the `toString` obligation and then not check the else expression.

Reply:
This is a valid concern; will look into it

Reply:
I added the test in 4c3789d, and it works.  The reason it works is that the `@MustCall` type for the local variable is computed by the Must Call Checker, which correctly computes the LUB for the ternary expression as `@MustCall({"toString","clone"})`.  So `MustCallInvokedChecker` indeed finds the issue.

I think there is a way to clean up the above code more, by just storing the relevant (sub-)expression on the RHS of an assignment in the `LocalVarWithTree` object rather than the whole assignment.  Then we might be able to get rid of all the counting logic.  But that is a broader change we can look into in a follow-up PR.  I'll open an issue on that once this lands.